packages/app/src/systems/DApp/machines/addAssetMachine.tsx

```diff
@@ -0,0 +1,141 @@
+import type { Asset } from '@fuel-wallet/types';
+import type { InterpreterFrom, StateFrom } from 'xstate';
+import { assign, createMachine } from 'xstate';
+
+import type { AssetInputs } from '~/systems/Asset';
+import { AssetService } from '~/systems/Asset';
+import { assignErrorMessage, FetchMachine } from '~/systems/Core';
+
+type MachineContext = {
+  asset?: Asset;
+  origin?: string;
+  error?: string;
+  addedAsset?: Asset;
+};
+
+type MachineServices = {
+  saveAsset: {
+    data: Asset;
+  };
+};
+
+type MachineEvents =
+  | {
+      type: 'START_ADD_ASSET';
+      input: { origin: string; asset: Asset };
+    }
+  | { type: 'ADD_ASSET' }
+  | { type: 'REJECT' };
+
+export const addAssetMachine = createMachine(
+  {
+    predictableActionArguments: true,
+    // eslint-disable-next-line @typescript-eslint/consistent-type-imports
+    tsTypes: {} as import('./addAssetMachine.typegen').Typegen0,
+    schema: {
+      context: {} as MachineContext,
+      services: {} as MachineServices,
+      events: {} as MachineEvents,
+    },
+    id: '(machine)',
+    initial: 'idle',
+    states: {
+      idle: {
+        on: {
+          START_ADD_ASSET: {
+            actions: ['assignAssetData'],
+            target: 'reviewAsset',
+          },
+        },
+        after: {
+          /** connection should start quickly, if not, it's probably an error or reloading.
+           * to avoid stuck black screen, should close the window and let user retry */
+          TIMEOUT: '#(machine).closing', // retry
+        },
+      },
+      closing: {
+        entry: ['closeWindow'],
+        always: {
+          target: '#(machine).failed',
+        },
+      },
```

Comment:
This is not required anymore, we have implemented a closePopUp when the connection is lost (user refresh)